src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -168,163 +353,192 @@ private BoundStatement RewriteEnumeratorForEachStatement(BoundForEachStatement n
             MethodSymbol disposeMethod;
             if (enumeratorInfo.NeedsDisposeMethod && Binder.TryGetSpecialTypeMember(_compilation, SpecialMember.System_IDisposable__Dispose, forEachSyntax, _diagnostics, out disposeMethod))
             {
-                Binder.ReportDiagnosticsIfObsolete(_diagnostics, disposeMethod, forEachSyntax,
-                                                   hasBaseReceiver: false,
-                                                   containingMember: _factory.CurrentMethod,
-                                                   containingType: _factory.CurrentType,
-                                                   location: enumeratorInfo.Location);
+                BoundStatement tryFinally = WrapWithTryFinallyDispose(forEachSyntax, enumeratorInfo, enumeratorType, boundEnumeratorVar, whileLoop, disposeMethod, disposeAwaitableInfoOpt: null);
```

Comment:
>WrapWithTryFinallyDispose [](start = 44, length = 25)

:memo: The `WrapWithTryFinallyDispose` was extracted from existing code and updated to handle the async case (it puts an extra `await` in that case).